component/loki/relabel/relabel.go

```diff
@@ -0,0 +1,214 @@
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/common/loki"
+	flow_relabel "github.com/grafana/agent/component/common/relabel"
+	lru "github.com/hashicorp/golang-lru"
+	prometheus_client "github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "loki.relabel",
+		Args:    Arguments{},
+		Exports: Exports{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments holds values which are used to configure the loki.relabel
+// component.
+type Arguments struct {
+	// Where the relabelled metrics should be forwarded to.
+	ForwardTo []loki.LogsReceiver `river:"forward_to,attr"`
+
+	// The relabelling rules to apply to each log entry before it's forwarded.
+	RelabelConfigs []*flow_relabel.Config `river:"rule,block,optional"`
+}
+
+// Exports holds values which are exported by the loki.relabel component.
+type Exports struct {
+	Receiver loki.LogsReceiver `river:"receiver,attr"`
+}
+
+// Component implements the loki.relabel component.
+type Component struct {
+	mut      sync.RWMutex
+	opts     component.Options
+	rcs      []*relabel.Config
+	receiver loki.LogsReceiver
+	fanout   []loki.LogsReceiver
+
+	entriesProcessed prometheus_client.Counter
+	entriesOutgoing  prometheus_client.Counter
+	cacheHits        prometheus_client.Counter
+	cacheMisses      prometheus_client.Counter
+	cacheSize        prometheus_client.Gauge
+
+	cache *lru.Cache
+}
+
+var (
+	_ component.Component = (*Component)(nil)
+
+	maxCacheSize = 10_000 // TODO(@tpaschalis) Do we make this configurable?
+)
+
+// New creates a new loki.relabel component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	cache, err := lru.New(maxCacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Component{
+		opts:  o,
+		cache: cache,
+	}
+	c.entriesProcessed = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_entries_processed",
+		Help: "Total number of log entries processed",
+	})
+	c.entriesOutgoing = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_entries_written",
+		Help: "Total number of log entries forwarded",
+	})
+	c.cacheMisses = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_cache_misses",
+		Help: "Total number of cache misses",
+	})
+	c.cacheHits = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_cache_hits",
+		Help: "Total number of cache hits",
+	})
+	c.cacheSize = prometheus_client.NewGauge(prometheus_client.GaugeOpts{
+		Name: "loki_relabel_cache_size",
+		Help: "Total size of relabel cache",
+	})
```

Comment:
nit: to clean this up a little bit, what do you think about declaring a metrics struct and putting it & these in metrics.go? That's a pattern we sometimes use in the codebase.

Reply:
You're right, I just bluntly copy/pasted from another component. Let me tidy this up :)